### PR TITLE
feat(lerna):  use docker and npx

### PR DESCRIPTION
**Pull request by viceice**
ref: renovatebot/github-action#9

### CODE REVIEW

**Comment by viceice** [comment]
semantic release dry-run failed again. don't really know why. Otherwise ready to review.

**Review by JamieMagee — DISMISSED** [review]

File: lib/manager/npm/post-update/lerna.ts [hunk 1]
```diff
@@ -44,10 +46,31 @@ export async function generateLockFiles(
       params =
         '--ignore-scripts --ignore-engines --ignore-platform --mutex network:31879';
     }
-    cmd = `npm i -g -C ~/.npm/lerna@${lernaVersion} lerna@${lernaVersion} && ${lernaClient} install ${params} && ~/.npm/lerna@${lernaVersion}/bin/lerna bootstrap --no-ci -- ${params}`;
-    if (binarySource === BinarySource.Global) {
-      cmd = `${lernaClient} install ${params} && lerna bootstrap --no-ci -- ${params}`;
+    if (config.binarySource === BinarySource.Docker) {
```
Comment (rarkins) [inline lib/manager/npm/post-update/lerna.ts:49]: Can we migrate this to use lib/util/exec ?

Comment (viceice) [reply lib/manager/npm/post-update/lerna.ts:49]: Sure

Comment (viceice) [reply lib/manager/npm/post-update/lerna.ts:49]: Was copy and paste from yarn / npm 🙈

Comment (viceice) [reply lib/manager/npm/post-update/lerna.ts]: done

**Comment by rarkins** [comment]
I want to do some real world tests on this first, so trying with a fork of `jest` here: https://github.com/renovate-tests/jest/pulls

Starting with the implementation in `master` and then will switch.

**Comment by rarkins** [comment]
@viceice it looks like our existing `lerna` implementation may need updating. IIRC we currently run `lerna bootstrap` instead of `yarn` "just in case". i.e. I wasn't sure if it was safe to assume `yarn` would do everything. However it seems `lerna` made an update to force you to run `yarn` if workspaces are configured. e.g. this error:

```
$ lerna bootstrap --npm-client=yarn
lerna info version 2.9.0
lerna ERR! EWORKSPACES Yarn workspaces are configured in package.json, but not enabled in lerna.json!
lerna ERR! EWORKSPACES Please choose one: useWorkspaces = true in lerna.json, or remove package.json workspaces config
lerna ERR! initialize callback with error
lerna ERR! ValidationError: Yarn workspaces are configured in package.json, but not enabled in lerna.json!
lerna ERR! Please choose one: useWorkspaces = true in lerna.json, or remove package.json workspaces config
lerna ERR!     at new ValidationError (/Users/greada/.config/yarn/global/node_modules/lerna/lib/utils/ValidationError.js:25:120)
lerna ERR!     at BootstrapCommand.initialize (/Users/greada/.config/yarn/global/node_modules/lerna/lib/commands/BootstrapCommand.js:138:25)
```

I think we need to fix this logic in our master first before it's possible to properly verify this branch. Note: this change may mean that repos like `jest` then look and behave just like a regular `yarn` one and we don't run any specific `lerna` commands.

**Comment by viceice** [comment]
Ok, I'll have a look later. Maybe we need a new version check and /or some properties check's.

**Comment by rarkins** [comment]
Possibly we can keep calling `lerna` but include [`--use-workspaces`](https://github.com/lerna/lerna/tree/master/commands/bootstrap#--use-workspaces). Looking at Jest's `package.json` I think they maybe run `yarn` instead though - there's no `lerna bootstrap` shortcut in the repo's scripts.

**Comment by rarkins** [comment]
Indeed, their [contributing guide](https://github.com/facebook/jest/blob/master/CONTRIBUTING.md) says to run `yarn`.

**Comment by rarkins** [comment]
Another alternative: find a project that uses lerna + npm and test this PR against that.

**Comment by viceice** [comment]
They have an invalid lerna version [25.3.0](https://github.com/renovate-tests/jest/blob/f62d85e4340dac3c8e33dadf1f2908a8f4e77aa9/lerna.json#L2)

**Comment by rarkins** [comment]
Verified `npx` working ok in this (basic) lerna/npm example: https://github.com/renovate-tests/lerna-monorepo-example/pull/6/files

**Comment by rarkins** [comment]
Is it possible to raise a new PR that only changes the global lerna to use `npx`? i.e. do not add binarySource=docker awareness yet. Doing so could actually break something in the app.

**Comment by viceice** [comment]
ok

**Comment by viceice** [comment]
done

**Comment by viceice** [comment]
done

**Pull request merged**